portmidi/update.sh

```diff
@@ -1,62 +1,119 @@
-#! /bin/bash
+#! /bin/sh
 #
 # this script automatically downloads and copies the portmidi library for Pd
 #
-# uses svn trunk version by default, set checkout revision # using
+# uses git master version by default, set checkout branch/tag/commit using
 # first argument:
 #
-#    ./update.sh 239
+#    ./update.sh c0d239712d
 #
-# Dan Wilcox danomatika.com 2018
+# Dan Wilcox danomatika.com 2018,2024
 #
 
 # exit on error
 set -e
 
 ##### VARIABLES
 
-SRC=portmidi-svn
+SRC=portmidi-git
 DEST=portmidi
 
-# commandline arg: version is svn revision #
+# commandline arg: version can be a tag, commit, etc
+# uses git master by default
 VERSION=""
 
 ##### FUNCTIONS
 
 # copy .h & .c files from $SRC to $DEST, ignore missing file errors
 function copysrc {
-    mkdir -p $DEST/$1
-    cp -v $SRC/$1/*.h $DEST/$1 2>/dev/null || :
-    cp -v $SRC/$1/*.c $DEST/$1 2>/dev/null || :
+    mkdir -p "${DEST}/${1}"
+    cp -v "${SRC}/${1}"/*.h "${DEST}/${1}" 2>/dev/null || :
+    cp -v "${SRC}/${1}"/*.c "${DEST}/${1}" 2>/dev/null || :
 }
 
-##### GO
+usage() {
+cat <<EOF 1>&2
+Usage: $0 [OPTIONS] [VERSION]
+
+Options:
+  -h       display this help message
+
+  -f       force (do not abort if '${DEST}/' exists)
+           this updates existing files but keeps removed files,
+           only use this if you know what you are doing!
+
+Arguments:
+
+VERSION    tag/commit/branch of portaudio to import
+EOF
+}
+
+#### ARGUMENTS
+
+force=no
+OPTIND=1
+while getopts ":hf" opt; do
+    case "$opt" in
+        h)
+            usage
+            exit 0
+            ;;
+        f)
+            force=yes
+            ;;
+        '?')
+            usage
+            exit 1
+            ;;
+    esac
+done
+shift "$((OPTIND-1))" # shift off the options and optional --
 
-# append revision number to checkout rule ala @###
 if [ $# -gt 0 ] ; then
-    VERSION="@$1"
+    VERSION="$1"
 fi
 
+##### GO
+
 # move to this scripts dir
-cd $(dirname $0)
+scriptdir=$(dirname "$0")
+cd "${scriptdir}"
+
+# check if target exists
+if [ -d "${DEST}" ]; then
+    echo "Destination '${DEST}' exists" 1>&2
+    if [ "${force}" = "yes" ]; then
+        echo "  forced to update existing files" 1>&2
+    else
+        echo "  Please remove '$(pwd)/${DEST}' before running $0" 1>&2
+        echo "  or check the '-h' flag" 1>&2
+        exit 1
+    fi
+fi
 
-# checkout source
+# clone source
 echo "==== Downloading portmidi $VERSION"
-if [ -d $SRC ] ; then
-    rm -rf $SRC
+if [ -d "${SRC}" ] ; then
+    rm -rf "${SRC}"
+fi
+git clone https://github.com/PortMidi/portmidi.git "${SRC}"
```

Comment:
would it make sense for this to be --depth 1? (I guess it would be combined w/ 'checkout' in that case..)

Reply:
Possibly, but how do you then checkout an older tag if you only do a shallow clone?

Reply:
hmm I guess you can't..